File: Dynamic Programming/18427번 - 함께 블록 쌓기/18427.py
import sys

input = sys.stdin.readline

n, m, h = map(int, input().split())
students = [0] + [list(map(int, input().split()))for _ in range(n)]
dp = [[0] * (h + 1) for _ in range(n + 1)]

def solve():
    for i in range(n + 1):
        dp[i][0] = 1

    
    for i in range(1, n + 1):
        for j in range(1, h + 1):
            for k in students[i]:
                if j >= k:
                    dp[i][j] += dp[i - 1][j - k]
                    dp[i][j] %= 10007
                    
            dp[i][j] += dp[i - 1][j]
            dp[i][j] %= 10007

    print(dp[n][h])    
    
if __name__ == '__main__':
    solve()

File: Dynamic Programming/18427번 - 함께 블록 쌓기/18427.cpp
#include <iostream>
#include <vector>
#include <sstream>

using namespace std;

int n, m, h;
vector<int> students[51];
int dp[51][1001] = { 0, };
int ans = 0;

void input() {
    cin >> n >> m >> h;
    cin.ignore();
    
    for(int i = 1; i <= n; i++) {
        string tmp;
        getline(cin, tmp);
        
        istringstream iss(tmp);
        string buffer;
        
        while(getline(iss, buffer, ' ')) {
            students[i].push_back(stoi(buffer));
        }
    }    
}

void solve() {
    // i번째 학생이 높이 0을 만드는 경우의 수
    for(int i = 0; i <= n; i++) dp[i][0] = 1;
    
    for(int i = 1; i <= n; i++) {
        // i번째 학생이 쌓을 수 있는 높이들
        for(int j = 1; j <= h; j++) {
            // i번째 학생이 가진 블럭들
            for(auto &k : students[i]) {
                if(j >= k) {
                    dp[i][j] += dp[i - 1][j - k];
                    dp[i][j] %= 10007;
                }
            }
            
            // i번째 학생이 블럭을 사용안해도 가능한 경우의 수도 더함
            dp[i][j] += dp[i - 1][j];
            dp[i][j] %= 10007;
        }
    }
    
    cout << dp[n][h] << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}